# Segmentation

1) Donwload wiki.txt according to instructions, place it to C:\input in my PC

2) Import necessary libraries

In [1]:
from nltk.tokenize import sent_tokenize
import random

3) Extract 50 random non empty lines from wiki.txt and segment them with nltk

In [8]:
with open('C://mcl2018_hw1_babakov//practicals//Segmentation//wiki.txt','r',encoding = 'utf-8') as wiki:
    data = wiki.read();
    lines = data.splitlines()
    non_empty_lines = []
    for line in lines:#filter all empty lines
        if(line):
            non_empty_lines.append(line)
    tokenized_lines = []
    random.seed(42)
    rand_lines = random.sample(non_empty_lines, 50)
    for line in rand_lines:
        tokenized_line = sent_tokenize(line)
        tokenized_lines.append(tokenized_line)
    sep_sentences = 0
    print_count = 0 
    
    for tok_line in tokenized_lines:
        sep_sentences += len(tok_line)
        if (print_count < 15):
            print (tok_line, '\n')
            print_count += 1
    print("segmented sentences: ",sep_sentences)

['Согласно Конституции, законодательную власть осуществляет Федеральное собрание\xa0— парламент, состоящий из двух палат Совета Федерации и Государственной думы, в научной литературе данные палаты называются верхней и нижней соответственно.', 'В Совет Федерации входят по два представителя от каждого субъекта федерации.', 'В 1993 году члены Совета Федерации были избраны населением, в 1995 году их место заняли непосредственно губернаторы (президенты) и спикеры региональных законодательных собраний, а с 2000 года\xa0— назначаемые и избираемые представители соответственно исполнительного и законодательного органов власти субъектов федерации.', 'Государственная дума состоит из 450 депутатов, избираемых всенародным голосованием по партийным спискам сроком на 5 лет (до 31 декабря 2008 года\xa0— на 4 года).'] 

['Литва\xa0— парламентская республика, с характерными чертами президентской республики.', 'Срок полномочий избираемого всенародно Президента\xa0— 5 лет.', 'На данный момент президентом 

Examples of bad results from nltk

'Но одновременно быстрыми темпами шёл рост местного населения (см.',

'ниже\xa0— агломерация).', 

'Всё это начало создавать большую антропогенную нагрузку на экологию КМВ (см.', 'Пятигорск).'

In [14]:
with open("C:\mcl2018_hw1_babakov\practicals\Segmentation\wiki_50.txt", 'w',encoding = 'utf-8') as f:
    for t in tokenized_lines:
        f.write(''.join(str(s) for s in t) + '\n')

4) Run ruby script for segmentation, output results as output_ruby. Calculate number of segmented lines 

In [10]:
with open('C://mcl2018_hw1_babakov//practicals//Segmentation//output_ruby.txt','r',encoding = 'utf-8') as ruby:
    data = ruby.read();
    lines = data.splitlines()
    print("Number of sentences segmented by ruby is ",len(lines))

Number of sentences segmented by ruby is  177


Examples of bad results from ruby script

Вот почему 27 марта 1992 года и был подписан Указ Президента Российской Федерации Б.

 Н.
 
 Ельцина, согласно которому Кавминводы являются особо охраняемым эколого-курортным регионом Российской Федерации.

# Conclusion

I have calculated real amount of sentences manually. It equals 153

NLTK segmentation turned to be more accurate becuase it made only 164 - 153 = 11 non existing sentences

Meanwhile python script made 177 - 153 = 24 non existing sentences

NLTK main mistakes were about abbreviations of short words
Ruby scrip main mistkes were about name abbreviation